textbook/chapter_1.qmd

```diff
@@ -0,0 +1,993 @@
+2. **Geometric marks**: Visual elements representing data (e.g., bars, dots).
+3. **Visual encoding**: Rules for mapping data to visual properties.
+
+Unlike traditional charting tools that offer pre-made chart types, a grammar-based approach is more flexible. It allows the user to construct custom visualizations from fundamental building blocks.
+
+The following table illustrates the similarities between the grammars of programming, natural language and graphics:
+
+| Component | Natural Language | Programming Language | Graphical Grammar |
+|-----------|-------------------|----------------------|-------------------|
+| Basic Units | Words | Variables, literals | Marks (points, lines, bars) |
+| Syntax Rules | Word order, punctuation | Syntax rules, operators | Rules for combining marks, layers, aesthetics |
+| Semantics | Word meanings, context | Expression meanings | Meanings of visual encodings |
+| Composition | Sentences, paragraphs | Statements, code blocks | Layers, facets |
+| Output | Text | Executable code | Data visualizations |
+| Modifiers | Adjectives, adverbs | Attributes, parameters | Scales, legends, annotations |
+
+Understanding these parallels enables users to approach data visualization systematically, leveraging a defined set of components and rules.
+
+### Coding with grammar in Altair and ggplot
+
+Altair and ggplot offer a powerful and concise visualization grammar for quickly building a wide range of statistical graphics. 
+
+In brief, you first create a canvas/chart, then you encode your data variables as different dimensions in this chart (x, y, color, etc) and add geometric marks to represent the data (points, lines, etc). You can see an illustration of this in the image below, together with how `ggplot` (top) and `altair` (bottom) uses the grammatical components in its syntax. The exact syntax is slightly different, but the overall structure very much the same:
+
+![Alt text](/img/grammar-of-graphics.jpeg)
+
+**Let's get coding**
+
+Data in Altair and ggplot is built around ["tidy"](http://vita.had.co.nz/papers/tidy-data.html) dataframes,
+which consists of a set of named data *columns*
+with one feature each and rows with one observation each.
+We will also regularly refer to data columns as data *fields*.
+We will focus on Altair but you can toggle codeblocks to see how to create the same charts in ggplot.
+
+
+We will often use datasets
+from the [vega-datasets](https://github.com/vega/vega-datasets) repository, some which are directly available as Pandas data frames. The cars dataset provides a rich opportunity to explore historical trends in automotive design, performance, and efficiency.
+
+::: {.panel-tabset}
+## Altair
+```{python}
+import altair as alt
+import pandas as pd
+
+
+url = 'https://raw.githubusercontent.com/joelostblom/teaching-datasets/main/cars.csv'
+# `parse_dates` makes sure a column is stored as a date object in pandas
+cars = pd.read_csv(url, parse_dates=['Year'])
+cars
+```
```

Comment:
I think the tabs work well here, but maybe we want to set the pandas options differently so that fewer rows are shown by default. Not sure.